go/libkb/key_pseudonym.go

```diff
@@ -0,0 +1,286 @@
+// (the key is referred to by the application id, the generation number and the user or team id).
+// Pseudonyms are only used for saltpack encryption at the moment, but more applications are possible. They are used
+// to avoid that, when decrypting a message encrypted with a team key, a user has to loop through all the keys of all the
+// teams he is part of to find the right one. To avoid this, when encrypting for a team, the sender generates such a pseudonym
+// (which refers to the key it used to encrypt), sends it to the server and includes it as a recipient identifier for the
+// appropriare recipient payload box in the header of the saltpack message. When decrypting a saltpack message, a recipient
+// can ask the server if any of the recipient identifiers in the message correspond to known pseudonyms (for teams the user is part of),
+// and use the response from the server to identify which key to use for decryption. This mechanism substitutes an older kbfs based one
+// (which is still supported to allow decryptions of old saltpack messages).
+//
+// A pseudonym is computed as an HMAC (with a random nonce as a key) of the information it refers to in order to
+// prevent the server from tampering with the mapping (we rely on collision resistance and not unforgeability,
+// as the server knows the nonce, so a simple SHA256 would have worked as well).
+type KeyPseudonym [32]byte
+
+type KeyPseudonymNonce [32]byte
+
+const KeyPseudonymVersion = 1
+
+// KeyPseudonymInfo contains the KeyPseudonym as well as information about the key it represents.
+type KeyPseudonymInfo struct {
+	KeyPseudonym KeyPseudonym
+	ID           keybase1.UserOrTeamID
+	Application  keybase1.TeamApplication
+	KeyGen       KeyGen
+	Nonce        KeyPseudonymNonce
+}
+
+func (t KeyPseudonym) String() string {
+	return hex.EncodeToString(t[:])
+}
+
+func (t KeyPseudonym) Eq(r KeyPseudonym) bool {
+	return hmac.Equal(t[:], r[:])
+}
+
+// keyPseudonymReq is what the server needs to store a pseudonym
+type keyPseudonymReq struct {
+	Pseudonym   string `json:"key_pseudonym"`   // hex
+	ID          string `json:"user_or_team_id"` // hex
+	Application int    `json:"app_id"`
+	KeyGen      int    `json:"key_gen"`
+	Nonce       string `json:"nonce"` // hex
+}
+
+// keyPseudonymContents is the data packed inside the HMAC
+type keyPseudonymContents struct {
+	_struct     bool `codec:",toarray"`
+	Version     int
+	ID          [16]byte                 // keybase1.UserOrTeamId as a byte array
+	Application keybase1.TeamApplication // int
+	KeyGen      KeyGen                   // int
+}
+
+type getKeyPseudonymsRes struct {
+	KeyPseudonyms []getKeyPseudonymRes `json:"key_pseudonyms"`
+	Status        AppStatus            `json:"status"`
+}
+
+func (r *getKeyPseudonymsRes) GetAppStatus() *AppStatus {
+	return &r.Status
+}
+
+type getKeyPseudonymRes struct {
+	Err  *PseudonymGetError `json:"err"`
+	Info *struct {
+		ID          string `json:"user_or_team_id"` // hex
```

Comment:
if you use the appropriate AVDL types here, you can get the decoder to unmarshal for you.

that is, make this `keybase1.UserOrTeamID` and then  make a `func (i *UserOrTeamID) UnmarshalJSON(b []byte) error` function.